File: leetcode/sumOfDistancesInTree.java
/**********************************************
  > File Name		: sumOfDistancesInTree.java
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Wed 15 Dec 2021 08:26:50 PM CST
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

class Solution {
    public int[] sumOfDistancesInTree(int N, int[][] edges) {
        int[] result = new int[N];
        ArrayList<Node> nodeList = new ArrayList<Node>(N);
        for (int i = 0; i < N; i++) {
            nodeList.add(new Node(i));
        }
        for (int i = 0; i < edges.length; i++) {
            int a = edges[i][0];
            int b = edges[i][1];
            Node nodeA = nodeList.get(a);
            Node nodeB = nodeList.get(b);
            nodeA.link(nodeB);
        }


        Node root = nodeList.get(0);
        root.traverse();//build tree
        root.initiate();
        root.totalDistance = root.subDistance;
        root.calculateDistance(N);
        for (int i = 0; i < N; i++) {
            result[i] = nodeList.get(i).totalDistance;
        }

        
        return result;
    }

    class Node {
        int val;
        Node parent;
        List<Node> nodes;
        int size;
        boolean visited;
        int subDistance;
        int totalDistance;
        public Node(int val) {
            this.val = val;
            this.parent = null;
            this.nodes = new ArrayList<Node>();
            this.size = 1;
            this.visited = false;
            this.subDistance = 0;
            this.totalDistance = 0;
        }

        public void link(Node anotherNode) {
            this.nodes.add(anotherNode);
            anotherNode.nodes.add(this);
        }

        public void traverse() {
            // System.out.println("A");
            for (Node node: nodes) {
                node.parent = this;
                node.nodes.remove(this);
                node.traverse();
            }
        }

        public void initiate() {
            for (Node node: nodes) {
                node.initiate();
                this.size += node.size;
                this.subDistance += node.subDistance;//sum of distances to all child nodes.
                this.subDistance += node.size;
            }
        }

        public void calculateDistance(int N) {
            for (Node node: nodes) {
                node.totalDistance = this.totalDistance + N - node.size * 2;
                //node.subDistance: current node to all it's child nodes distances.
                //this.totalDistance - node.size - node.subDistance: parent distances from all other child nodes.
                //N - node.size: current node to all non child nodes distances.
                //node.totalDistance = node.subDistance + this.totalDistance - node.size - node.subDistance + N - node.size = this.totalDistance + N - 2*node.size;
                node.calculateDistance(N);
            }
        }

        // public void traverse() {
        //     // int count = 0;
        //     // System.out.println(this.val);
        //     this.visited = true;
        //     for (Node node: nodes) {
        //         if (!node.visited) {
        //             node.distance = this.distance + 1;
        //             node.traverse();
        //         }
        //     }
        // }

    }
}

File: leetcode/sumOfDistancesInTree.cpp
/**********************************************
  > File Name		: sumOfDistanceInTree.cpp
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Tue 14 Dec 2021 08:09:47 PM CST
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

#include <vector>
#include <stdio.h>
#include <cassert>
#include <unistd.h>
#include "printv.h"
using namespace std;

class Solution {
public:
    vector<int> sumOfDistancesInTree(int n, vector<vector<int>>& edges) {

    }

    //time limit exceeded
    vector<int> sumOfDistancesInTree2(int n, vector<vector<int>>& edges) {
        vector<vector<int>> dists;
        
        for (int i = 0; i < n; i++) {
            dists.push_back(vector<int>(n, n));
            dists[i][i] = 0;
        }

        int count = n*(n-1) - 2*edges.size();
        for (vector<int>& edge: edges) {
            dists[edge[0]][edge[1]] = 1;
            dists[edge[1]][edge[0]] = 1;
        }

        vector<bool> done(n, false);
        int i, j, k;
        bool flag;
        int step = 2, temp;
        while (count) {
            for (i = 0; i < n; i++) {
                if (done[i]) {
                    continue;
                }
                flag = true;
                for (j = 0; j < n; j++) {
                    if (dists[i][j] > step) {
                        for (k = 0; k < n; k++) {
                            temp = dists[i][k] + dists[k][j];
                            if (temp <= step) {
                                dists[i][j] = temp;
                                dists[j][i] = temp;
                                count -= 2;
                                break;
                            }
                        }
                        if (dists[i][j] >= step) {
                            flag = false;
                        }
                    }
                }
                done[i] = flag;
                //printf("node %d, count = %d\n", i, count);
                //printv2(dists);
                //printf("\n");
                //sleep(1);

                if (!count) {
                    break;
                }
            }
            step++;
        }
        
        vector<int> res(n, 0);
        for (int i = 0; i < n; i++) {
            vector<int>& v = dists[i];
            for (int dist: v) {
                assert(dist < n);
                res[i] += dist;
            }
        }
        return res;
    }

};

int main() {
    Solution s;
    vector<vector<int>> edges = {{0,1},{0,2},{2,3},{2,4},{2,5}};
    //vector<vector<int>> edges = {{1,0}};
    vector<int> v = s.sumOfDistancesInTree(6, edges);
    printv(v);
}
